File: rungrip.py
import grip
import cv2
import numpy
import math
import config

def getContours(image):
	gripfunc = grip.GripPipeline()
	gripfunc.process(image)
	contours = gripfunc.filter_contours_output
	return contours;

def filterContours(image, contours):
	cnt = contours[0]
	for n in contours:
		if cv2.contourArea(n) >= cv2.contourArea(cnt):
			cnt = n
	if config.display:
		cv2.drawContours(image, cnt, -1, (0,0,0), 7)
	return cnt

def centroid(contour, image):
	M = cv2.moments(contour)
	# print M
	cx = int(M['m10']/M['m00'])
	cy = int(M['m01']/M['m00'])
	cv2.circle(image, (cx, cy), 15, [25,25,245], -1)
	return (cx, cy)

def rectangle(contour, image, centroid):
	rect = cv2.minAreaRect(contour)
	rectwidth = int(rect[1][1])
	rectlength = int(rect[1][0])
	angledeg = 90 - rect[2]
	if angledeg > 90:
		angledeg = angledeg - 180
	if rectlength > rectwidth:
		angledeg = angledeg + 90
	anglerad = numpy.radians(angledeg)
	if config.display:
		cx = centroid[0]
		cy = centroid[1]
		box = cv2.boxPoints(rect)
		box = numpy.int0(box)
		img = cv2.drawContours(image,[box],0,(25,25,245),8)
		if rectwidth > rectlength:
			img = cv2.line(image, (int(cx + math.cos(anglerad)*rectwidth/2), int(cy - math.sin(anglerad)*rectwidth/2)), (int(cx - math.cos(anglerad)*rectwidth/2), int(cy + math.sin(anglerad)*rectwidth/2)), (0,0,0), 8)
		else:
			img = cv2.line(image, (int(cx + math.cos(anglerad)*rectlength/2), int(cy - math.sin(anglerad)*rectlength/2)), (int(cx - math.cos(anglerad)*rectlength/2), int(cy + math.sin(anglerad)*rectlength/2)), (0,0,0), 8)
	return rectwidth/2, angledeg

File: runGrip.cpp
#include <vector>
#include <opencv2/opencv.hpp>
#include "config.h"
#include "GripPipeline.h"
#include "runGrip.h"

std::vector<std::vector<cv::Point>> RunGrip::getContours(cv::Mat &mat) {
    grip::GripPipeline pipeline;
    pipeline.Process(mat);
    return *pipeline.GetFilterContoursOutput();
}

std::vector<cv::Point> RunGrip::filterContours(std::vector<std::vector<cv::Point>> &contours) {
    double max_area = 0;
    std::vector<cv::Point> max_contour;
    for (std::vector<cv::Point> &c : contours) {
        double area = cv::contourArea(c);
        if (area > max_area) {
            max_area = area;
            max_contour = c;
        }
    }
    return max_contour;
}

cv::Point RunGrip::centroid(std::vector<cv::Point> &contour) {
    auto m = cv::moments(contour);
    return cv::Point(m.m10 / m.m00, m.m01 / m.m00);
}

double RunGrip::rectangle(cv::Mat &mat, std::vector<cv::Point> &contour, cv::Point &centroid) {
    cv::RotatedRect rect = cv::minAreaRect(contour);
    double width = rect.size.width;
    double length = rect.size.height;
    double angle = 90 - rect.angle;  // in degrees
    if (angle > 90) angle -= 180;
    if (length > width) angle += 90;
    if (DEBUG) {
        cv::drawContours(mat, std::vector<cv::RotatedRect> {rect}, 0, cv::Scalar(25, 25,  245), 8);
    }
    return angle;
}